core/coreunix/test/metadata_test.go

```diff
@@ -173,6 +174,20 @@ func addFileToBtfs(node *core.IpfsNode, data []byte, metadata string) (ipath.Pat
 		return nil, ctx.Err()
 	}
 
+	// The following adder.AddInProcess-for-loop mechanism is to wait for
+	// AddAllAndPin() to finish its callng of adder.PinRoot() since AddAllAndPin()
```

Comment:
The assumption that `AddAllAndPin` can return without finish calling `adder.PinRoot` is already a wrong assumption. `AddAllAndPin` should be a synchronous process where it should only return when all the sub-logic is done. So adding a workaround to wait for an async routine to finish here is a hack that should not be used.

Reply:
Checking the code. But there could be some reasons to do that. Without the metata tests, the current logic of AddAllAndPin works.

Reply:
Like progressReader, i don't have an option to check details.

Reply:
For the case of directory, AddAllAndPin() updates progressReader by sending file root node to the adder.Out channel. Then at the last line of the function, PinRoot() is to called to add the top node of the directory.